src/Microsoft.Tye.Core/ProjectReader.cs

```diff
@@ -30,94 +39,136 @@ public static async Task ReadProjectDetailsAsync(OutputContext output, FileInfo
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void EvaluateProject(OutputContext output, FileInfo projectFile, Project project)
         {
+            var sw = Stopwatch.StartNew();
+
+            ProjectInstance projectInstance;
             try
             {
-                output.WriteDebugLine("Installing msbuild targets.");
-                TargetInstaller.Install(projectFile.FullName);
-                output.WriteDebugLine("Installed msbuild targets.");
+                output.WriteDebugLine($"Loading project '{projectFile.FullName}'.");
+                projectInstance = ProjectInstance.FromFile(projectFile.FullName, new ProjectOptions());
+                output.WriteDebugLine($"Loaded project '{projectFile.FullName}'.");
             }
-            catch (Exception ex)
+            catch
             {
-                throw new CommandException("Failed to install targets.", ex);
+                throw new CommandException($"Failed to load project: '{projectFile.FullName}'.");
             }
 
-            var outputFilePath = Path.GetTempFileName();
+            // Currently we only log at debug level.
+            var logger = new ConsoleLogger(
+                verbosity: LoggerVerbosity.Normal,
+                write: message => output.WriteDebug(message),
+                colorSet: null,
+                colorReset: null);
 
             try
             {
-                var capture = output.Capture();
-                var programRoot = Path.GetDirectoryName(typeof(Program).Assembly.Location);
-
-                var restore = string.Empty;
-                if (!File.Exists(Path.Combine(projectFile.DirectoryName, "obj", "project.assets.json")))
-                {
-                    restore = "/restore";
-                }
-
-                output.WriteDebugLine("Running 'dotnet msbuild'.");
-                var msbuildCommand = $"msbuild {restore} /t:EvaluateTyeProjectInfo \"/p:TyeTargetLocation={programRoot}\" \"/p:TyeOutputFilePath={outputFilePath}\"";
-
-                output.WriteCommandLine("dotnet", msbuildCommand);
-                var exitCode = await Process.ExecuteAsync(
-                    $"dotnet",
-                    msbuildCommand,
-                    workingDir: projectFile.DirectoryName,
-                    stdOut: capture.StdOut,
-                    stdErr: capture.StdErr);
-
-                output.WriteDebugLine($"Done running 'dotnet msbuild' exit code: {exitCode}");
-                if (exitCode != 0)
-                {
-                    throw new CommandException("'dotnet msbuild' failed.");
-                }
+                AssemblyLoadContext.Default.Resolving += ResolveAssembly;
+                var result = projectInstance.Build(
+                    targets: new[] { "Restore", "ResolveReferences", "ResolvePackageDependenciesDesignTime", "PrepareResources", "GetAssemblyAttributes", },
```

Comment:
Running restore in the same operation as other things might cause some problems here. MSBuild 1) evaluates a project and then 2) runs targets on that evaluated project. When `Restore` happens, it may cause additional MSBuild logic to exist in places that would affect a subsequent evaluation--but it won't affect the current instance of the project (because that was already evaluated).

Looking at this list of targets, the concrete problems I'd expect are:
* Missing resources if they're pulled in through a package, and
* Incorrect assembly attributes if they're modified through a package (for example versioning through GitVersion or something).

MSBuild.exe implements `-restore` with 

https://github.com/microsoft/msbuild/blob/3453beee039fb6f5ccc54ac783ebeced31fec472/src/MSBuild/XMake.cs#L1427-L1438

It might be best to copy that approach here.

Reply:
We've already hit some of these issues, I'm going to try this 👍